instrumentation/tomcat/src/main/java/com/splunk/opentelemetry/instrumentation/tomcat/metrics/TomcatThreadPoolOtelMetrics.java

```diff
@@ -0,0 +1,73 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.instrumentation.tomcat.metrics;
+
+import com.splunk.opentelemetry.javaagent.bootstrap.metrics.otel.ThreadPoolMetrics;
+import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.api.OpenTelemetry;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.tomcat.util.net.AbstractEndpoint;
+
+final class TomcatThreadPoolOtelMetrics {
+  private static final String INSTRUMENTATION_NAME =
+      "com.splunk.javaagent.tomcat-thread-pool-metrics";
+
+  // a weak map does not make sense here because each Meter holds a reference to the executor
+  // AbstractEndpoint does not implement equals()/hashCode(), so it's safe to keep them in a plain
+  // ConcurrentHashMap
+  private static final Map<AbstractEndpoint<?, ?>, List<AutoCloseable>> threadPoolMetrics =
+      new ConcurrentHashMap<>();
+
+  public static void registerMetrics(AbstractEndpoint<?, ?> endpoint) {
+    threadPoolMetrics.computeIfAbsent(endpoint, TomcatThreadPoolOtelMetrics::createMeters);
+  }
+
+  private static List<AutoCloseable> createMeters(AbstractEndpoint<?, ?> endpoint) {
+    OpenTelemetry openTelemetry = GlobalOpenTelemetry.get();
+    ThreadPoolMetrics metrics =
+        ThreadPoolMetrics.create(openTelemetry, INSTRUMENTATION_NAME, "tomcat", endpoint.getName());
+
+    LateBindingTomcatThreadPoolMetrics tomcatThreadPoolMetrics =
+        new LateBindingTomcatThreadPoolMetrics(endpoint);
+    return Arrays.asList(
+        metrics.currentThreads(tomcatThreadPoolMetrics::getCurrentThreads),
+        metrics.activeThreads(tomcatThreadPoolMetrics::getActiveThreads),
+        metrics.idleThreads(tomcatThreadPoolMetrics::getIdleThreads),
+        metrics.coreThreads(tomcatThreadPoolMetrics::getCoreThreads),
+        metrics.maxThreads(tomcatThreadPoolMetrics::getMaxThreads),
+        metrics.tasksSubmitted(tomcatThreadPoolMetrics::getSubmittedTasks),
+        metrics.tasksCompleted(tomcatThreadPoolMetrics::getCompletedTasks));
+  }
+
+  public static void unregisterMetrics(AbstractEndpoint<?, ?> endpoint) {
+    List<AutoCloseable> observableInstruments = threadPoolMetrics.remove(endpoint);
+    if (observableInstruments != null) {
+      for (AutoCloseable observable : observableInstruments) {
+        try {
+          observable.close();
+        } catch (Exception exception) {
+          throw new IllegalStateException("Failed to close instrument", exception);
```

Comment:
This is because the common supertype for otel instruments is `AutoCloseable` which declares `close` as throwing exception while the `close` method in the instruments doesn't declare throwing exception. Here we catch the exception only because we are using `AutoCloseable`, we don't expect anything to be thrown.

Reply:
In my opinion you are overthinking this. By the same logic we should also do whatever we are doing here in `TomcatThreadPoolMicrometerMetrics.unregisterMetrics` because `Metrics.globalRegistry.remove` could have a bug that throws a runtime exception which could terminate the loop.
That being said I have changed rethrowing exception to logging a warning.